[TS] Add Obj API

khoitd1997 opened the pull request:
Hi, this is a PR to add obj based API for typescript. A couple of points:

- The PR won't build as is because I made a few updates to the [type file](https://github.com/khoitd1997/DefinitelyTyped/blob/patch-1/types/flatbuffers/index.d.ts), with the correct type file, all tests should pass
- If obj based API is enabled, the `flatbuffers` import need to be accessible for generated files because I need access to `createLong` for initializing the default values of class members
- I also modified the TS test runner to be more separate from the JS test runner for easier testing

khoitd1997 commented:
@googlebot I signed it!

aardappel commented on package-lock.json:
```diff
@@ -0,0 +1,13 @@
+{
```
should this file be in the repo?

aardappel commented on src/idl_gen_js_ts.cpp:
```diff
@@ -661,6 +674,486 @@ class JsTsGenerator : public BaseGenerator {
+  // $post_offset will include the value of variable declared in pre_offset
+  // as well as offset(or values) of things that can be serialized inline
+  // such as offset for struct or value for scalar
+  void FmtObjApi(const std::list<ObjApiData> &data, std::string &code,
```
Please generate code much like the code generation that already existed in this file, rather than inventing a new way of doing it based on regex, so that others can maintain this code easily.

khoitd1997 replied:
I changed it into a callback style. While I understand that inconsistency is bad for maintaining, it's hard to follow the exact current style because I would either need to add to the already big `GenStruct` function or repeat a loop multiple times to gather overlapping data. Do you have any suggestions?

aardappel replied:
My suggestion is to write the code much like the rest of the file was already doing, and all the other code generators are doing, meaning not first collecting parameters and formatting them afterwards, but doing it directly where needed, with no additional abstraction. That's most consistent and easiest to change for others. Adding code to an already big function is no concern.

khoitd1997 commented on js/flatbuffers.js:
```diff
@@ -830,6 +839,65 @@ flatbuffers.Builder.prototype.createString = function(s) {
+    return 0
+  }
+
+  if(typeof obj === 'string') {
```
This is here because it's also used for union types which can be `string`

khoitd1997 commented:
I can also do the `javascript` object API but I think this PR will go through many revisions so it's probably better to open another one after this one is merged.

krojew commented on tests/JavaScriptTest.js:
```diff
@@ -3,8 +3,12 @@ var assert = require('assert');
+
 var MyGame = require(process.argv[2]).MyGame;
 
+var isTsTest = process.env.FB_TS_TEST ? true : false; 
```
Use !! to cast to bool, rather than the ternary bool antipattern.

khoitd1997 commented on js/flatbuffers.js:
```diff
@@ -830,6 +839,49 @@ flatbuffers.Builder.prototype.createString = function(s) {
+  for(let i = 0; i < list.length; ++i) {
+    let val = list[i];
+
+    if(val !== null) {
```
I think this won't be an error because, in my [type file](https://github.com/khoitd1997/DefinitelyTyped/blob/patch-1/types/flatbuffers/index.d.ts), the member of the list can be null as well.

khoitd1997 commented:
@krojew I think I did have `--strict` because that's the default inside `TypeScriptTest.sh`

khoitd1997 commented:
@krojew @aardappel sorry for the delay, had to take some finals. I made the obj api code look more like the current code.

aardappel commented on src/idl_gen_js_ts.cpp:
```diff
@@ -661,6 +674,441 @@ class JsTsGenerator : public BaseGenerator {
+  std::string GetObjApiUnionTypeTS(const IDLOptions &opts,
+                                   const EnumDef &union_enum) {
+    std::string ret = "";
+    std::set<std::string> typeList;
```
please use `snake_case` for all variables, here and elsewhere

aardappel commented on tests/union_vector/union_vector_generated.ts:
```diff
@@ -426,4 +542,109 @@ static createMovie(builder:flatbuffers.Builder, mainCharacterType:Character, mai
+    let targetEnumStr = Character[this.mainCharacterType()];
+    if(targetEnumStr === 'NONE') { return null; } 
+
+    if(targetEnumStr === 'MuLan') { return (this.mainCharacter(new Attacker())! as Attacker).unpack();}
```
Does TS have a switch that works on strings?
Would it be good to stick this function in an actual named function? if a schema has multiple fields of the same union type, then all this code will be duplicated.

khoitd1997 commented on js/flatbuffers.js:
```diff
@@ -54,6 +54,15 @@ flatbuffers.FILE_IDENTIFIER_LENGTH = 4;
+ * @param {number} high
+ * @returns {flatbuffers.Long}
+ */
+flatbuffers.createLong = function(low, high) {
```
It's needed when creating default values for long types in the constructor of the object API. The ```GenDefaultValue``` function uses a hardcoded createLong method when creating long so exposing this function allows it to be compatible with that

khoitd1997 commented:
@aardappel I addressed most of the PR reviews, I also noticed that the generated TS code actually doesn't handle unions that have only strings, I added test cases as well as another union for the union_vector test, it results in most of the files of other languages being regenerated again too, is that okay?

aardappel commented:
It would be great to not add to the schema if possible, since it causes a lot of churn in other languages. Depends on wether you think this is essential for your functionality to not regress. We have no need need for a string only union in other languages, so I'd say no.

@krojew you otherwise happy with the above?

aardappel commented:
There's still CI errors:
```
In file included from /flatbuffers/tests/test.cpp:36:
/flatbuffers/tests/union_vector/union_vector_generated.h: In static member function 'static void* StringOnlyCharacterUnion::UnPack(const void*, StringOnlyCharacter, const resolver_function_t*)':
/flatbuffers/tests/union_vector/union_vector_generated.h:902:130: error: unused parameter 'resolver' [-Werror=unused-parameter]
 inline void *StringOnlyCharacterUnion::UnPack(const void *obj, StringOnlyCharacter type, const flatbuffers::resolver_function_t *resolver) {
                                                                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
/flatbuffers/tests/union_vector/union_vector_generated.h: In member function 'flatbuffers::Offset<void> StringOnlyCharacterUnion::Pack(flatbuffers::FlatBufferBuilder&, const rehasher_function_t*) const':
/flatbuffers/tests/union_vector/union_vector_generated.h:920:143: error: unused parameter '_rehasher' [-Werror=unused-parameter]
 inline flatbuffers::Offset<void> StringOnlyCharacterUnion::Pack(flatbuffers::FlatBufferBuilder &_fbb, const flatbuffers::rehasher_function_t *_rehasher) const {
```

krojew commented:
I'm generally happy with this 👍

khoitd1997 commented:
@aardappel @krojew all tests have passed, I have removed all of the string only union stuff.

aardappel commented:
Awesome, thanks!
